### DISCUSSION TITLE
Questions about boundary condition setting

### DISCUSSION THREAD

**Discussion by Joseph-0123** [idaholab/moose#23278] [Q&A General] [answered]
Now I have two systems that share one common interface.

Can I set same pressure and pressure gradient for their shared boundary?

**Comment by GiudGiud** [comment, marked as answer]
What variables do you have?
And are they defined on both sides of the interface, or one side each

**Reply by Joseph-0123** [reply]
> What variables do you have? And are they defined on both sides of the interface, or one side each

I have a fluid flow process with the Navier-Stoke flow (1st system) and Darcy flow (2nd system). In the 1st system, we have pressure and mass flow rate as variables and in the 2nd system, we have only pressure as variable.

So only pressure is defined on both sides but not others.

**Reply by GiudGiud** [reply]
what variable type is pressure?
is it continuous?

**Reply by Joseph-0123** [reply]
> what variable type is pressure? is it continuous?

pressure is continuous and I want to use multiApp to couple these two systems. pressure is scalar variable.

**Reply by Joseph-0123** [reply]
> what variable type is pressure? is it continuous?

So I want to compute variable values in one system then transfer to another system.

**Reply by GiudGiud** [reply]
A scalar variable is a single scalar for the entire mesh. 
Isn’t pressure a field variable instead?

Using a multi app approach, you can transfer a field. But here you need boundary values right?

**Reply by Joseph-0123** [reply]
> A scalar variable is a single scalar for the entire mesh. Isn’t pressure a field variable instead?
> 
> Using a multi app approach, you can transfer a field. But here you need boundary values right?

ok, pressure is field vairable in 3D grids. 
Yes, boundary values are not defined at the interface of two systems.

**Reply by GiudGiud** [reply]
I would replicate the mesh in both systems, and store the pressure from the OTHER system in an auxiliary variable.
Then the transfers will naturally map the pressure fields from each half of the mesh in one system to the same half on the other system. You may even be able to use the MultiAppCopyTransfer with block restriction

Then either:
- develop an interface kernel to handle the interaction between the two variables at the boundary
OR
- use a matched value BC to match the pressures

**Reply by Joseph-0123** [reply]
> I would replicate the mesh in both systems, and store the pressure from the OTHER system in an auxiliary variable. Then the transfers will naturally map the pressure fields from each half of the mesh in one system to the same half on the other system. You may even be able to use the MultiAppCopyTransfer with block restriction
> 
> Then either:
> 
> * develop an interface kernel to handle the interaction between the two variables at the boundary
>   OR
> * use a matched value BC to match the pressures
Thanks a lot.
I amnot sure if I understand - replicate the mesh in both system, Do you mean use one system containing these two sub-systems? Or one gmsh file contains both systems? Thanks again.

**Reply by Joseph-0123** [reply]
> I would replicate the mesh in both systems, and store the pressure from the OTHER system in an auxiliary variable. Then the transfers will naturally map the pressure fields from each half of the mesh in one system to the same half on the other system. You may even be able to use the MultiAppCopyTransfer with block restriction
> 
> Then either:
> 
> * develop an interface kernel to handle the interaction between the two variables at the boundary
>   OR
> * use a matched value BC to match the pressures

By the way, I have another question. I use our uni cluster to run simulations. But if I increase the numbers of the nodes and cores in the uni cluster, the 'Solve Did NOT Converge' will be shown in the outcome. What is the reason for this?

**Reply by GiudGiud** [reply]
Use one gmsh file in both applications. And use block restrictions to restrict the physics on both sides of this interface. 

The preconditioning is different in serial and in parallel.
You will want to find a preconditioner that works for you in parallel.

**Reply by Joseph-0123** [reply]
> Use one gmsh file in both applications. And use block restrictions to restrict the physics on both sides of this interface.
> 
> The preconditioning is different in serial and in parallel. You will want to find a preconditioner that works for you in parallel.

ok, I think I did exactly what you mentioned here. but one system is porous flow and another system is pipe flow, when I run the simulation, t requires a very small dt (0.0001s), or else it will not have convergence. umm, that is strange, huh?

**Reply by GiudGiud** [reply]
One of the two systems must be written incorrectly. Both solvers are implicit and can take large time steps. 

Please run only one at a time and look at the boundary conditions, the mesh, etc to find the issue.

**Reply by Joseph-0123** [reply]
> One of the two systems must be written incorrectly. Both solvers are implicit and can take large time steps.
> 
> Please run only one at a time and look at the boundary conditions, the mesh, etc to find the issue.

Thanks for your answer, I am trying to do that. 

By the way, I have another question: can we have different [Preconditioning] for different blocks? If so, do we have any examples in MOOSE ? Thanks.

**Reply by GiudGiud** [reply]
Like spatially?
I dont think so currently
